### ISSUE TITLE
Sign issue in Sloncewski spin torque

### ISSUE THREAD

**Issue by Gh0str1d3r** [mumax/3#99]
The Slonceswski spin torque in MuMax is defined as

tau = (A - alpha B) / (1+alpha^2) m x p x m + (B - alpha A)/(1+alpha^2) p x m

with

A = beta epsilon
B = beta epsilon_prime

I believe the first coefficient should be (A + alpha B). Please allow me to illustrate the issue.

The following python script determines the damping-like and field-like coefficients of the explicit LLGS, commonly called a_j and b_j, for a film thickness of 1 nm, Ms = 1e6, Lambda = 1. It just finds the field that one has to apply to keep the magnetization aligned in the out-of-plane direction.

```import subprocess
import numpy as np
import sys

def exec_command(command,verbose=False):
  '''
  Execute a command on the shell and return the standard output and
  the standard error as a result. The function will not return before
  the command has been executed and finished.
  
  Paramters
  ---------
  command : string
    Command (bash) to be executed
  verbose : bool
    It True, the command, the standard output and the standard error
    are printed to the terminal.
  
  Returns
  -------
  A tuple (standard output, standard error)
  '''
  if(verbose):  
    print(command)
  p = subprocess.Popen(
        command,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        shell=True)
  stdout, stderr = p.communicate()
  if ( stdout and verbose):
	  print('process returned stdout:\n' + stdout)
  if ( stderr and verbose):
	  print('process returned stderr:\n' + stderr)
  return stdout,stderr
  
script = """
SetGridSize(1, 1, 1)
SetCellSize(1e-6, 1e-6, 1e-9)
Msat = 1e6
Aex = 1e-11
alpha = 0.5
m = Uniform(0, 0, 1)
B_ext = vector({Bx}, {By}, 1)
DisableZhangLiTorque = True
EnableDemag = False
lambda = 1.0
Pol = {pol}
EpsilonPrime = {ep}
FixedLayer = vector(0.0, 1.0, 0)
OutputFormat = OVF1_TEXT
J = vector(0, 0, 1e12)
run(1e-9)
save(m)
"""


Bx=0
By=0
dx=0.1
dy=-0.1
signx=-1
signy=1
for i in range(20):
  with open("tmp.mx3", "w") as text_file:
    text_file.write(script.format(Bx=Bx,By=By,ep=sys.argv[1],pol=sys.argv[2]))
  command = "mumax3 tmp.mx3"
  exec_command(command)
  m = np.genfromtxt("tmp.out/m000000.ovf")
  print(Bx,By,m)
  if m[0]*signx < 0:
    dx /= -2
    signx *= -1
  if m[1]*signy < 0:
    dy /= -2
    signy *= -1
  if np.abs(m[1]) < 1e-4:
    dy = 0
  Bx += dx
  By += dy
  
print("SOT fields:")
print("mu_0 a_j = {:.3f} T".format(-Bx))
print("mu_0 b_j = {:.3f} T".format(-By))
 ```

The result for epsilon_prime = 0 (first parameter of the script, the second is the polarization)

`python determine_mumax_SOT.py 0 1`
> SOT fields:
> mu_0 a_j = -0.329 T
> mu_0 b_j = 0.000 T

is exactly what we expect for a film thickness of 1 nm, Ms = 1e6, Lambda = 1. However, changing epsilon_prime to 1/4 decreases the DL torque a_j, which it should not:

`python determine_mumax_SOT.py 0.25 1`
> SOT fields:
> mu_0 a_j = -0.197 T
> mu_0 b_j = 0.099 T`

This is not correct. We can achieve the correct result by scaling epsilon and epsilon_prime such that (A - alpha B) is actually (A + alpha B). This is achieved at a polarization of 5/3 and epsilon_prime = 5/12. Indeed, using these values, we get the expected result of an unchanged DL torque and a FL torque that is just half the DL torque (epsilon_prime = 1/4 corresponds to b_j = a_j / 2):

`python determine_mumax_SOT.py 0.41667 1.6667`
> SOT fields:
> mu_0 a_j = -0.329 T
> mu_0 b_j = 0.165 T

Last comment: I believe there is a factor gamma missing in the documentation (should be tau = gamma ...).

**Comment by ilemesh** [comment]
Good point, I have just double checked the derivation of the LL equation from LLG, and to me, it looks like the second coefficient in Gh0str1d3r's equation (the field-like torque) should be modified, while the first one kept intact. Also, the signs before the coefficients should be switched. See the proof below:
<img width="887" alt="screen shot 2017-03-14 at 10 19 02 pm" src="https://cloud.githubusercontent.com/assets/26417685/23922853/6f89f66a-0904-11e7-8546-9cfca2b6330e.png">

**Comment by barnex** [comment]
Has anyone modified the signs and checked whether standard problem 5a (https://github.com/mumax/3/blob/master/test/standardproblem5a.mx3) still passes? This test gives the same result as OOMMF.

**Comment by Gh0str1d3r** [comment]
Correct me if I'm wrong, but that standard problem 5a seems to not include the FL torque. epsilonprime = 0. The bug only pertains code with epsilonprime != 0.

**Comment by barnex** [comment]
OK, in that case I would suggest we need a new test with epsilonprime != 0 and golden values computed with OOMMF. The test should fail first and only pass after the signs have been corrected.

**Comment by Gh0str1d3r** [comment]
I can do this. I'm just not an expert on the code. Can you tell me which lines to edit and how to compile it to change the sign?

**Comment by barnex** [comment]
The implementation is here: https://github.com/mumax/3/blob/master/cuda/slonczewski2.cu .
Compile with ./make.bash.

Most important is to have the test file with values from OOMMF, however.

**Comment by Lizatre** [comment]
just curious, has this issue been fixed already for all the tests including slonczewski2?

**Comment by godsic** [comment]
Here is OOMMF's implementation.

@JeroenMulkers @JLeliaert Correct me if I am wrong but @Gh0str1d3r seems to be right
`     
      const OC_REAL8m cell_alpha = alpha[i];
      const OC_REAL8m cell_mgamma = -1*gamma[i]; // -1 * Landau-Lifshitz gamma

      ThreeVector p = mp[i];
      ThreeVector m = spin_[i];
      ThreeVector mxp = m; mxp ^= p;

      OC_REAL8m Bdot = B[i];
      if(!mp_propagate) {
        Bdot *= (m*p);
      }
      /// If mp_propagate is true, then mp[i] is actually something like
      /// (m[i+1]-m[i-1])/2, so m*p is in no sense cos(theta_m,p).
      /// Instead, we just take theta=0 so cos(theta_m_p)=1.  As an
      /// alternative, one could use sqrt(1-p*p), although this
      /// isn't right either if theta>pi/2.

      OC_REAL8m Aplus = A[i] + Bdot;
      OC_REAL8m Aminus = A[i] - Bdot;

      OC_REAL8m betaeps = beta_q_plus[i]/Aplus + beta_q_minus[i]/Aminus;
      OC_REAL8m betaepsp = beta_eps_prime[i];
      betaeps  *= Jmult*Ms_inverse_[i];
      betaepsp *= Jmult*Ms_inverse_[i];

      ThreeVector mxH = dm_dt_[i];

      OC_REAL8m coef1 = cell_alpha * betaeps;
      OC_REAL8m coef2 = cell_alpha * betaepsp;
      if(do_precess) {
        coef1 -= betaepsp;
        coef2 += betaeps;
      }

      ThreeVector scratchA = mxH;
      scratchA *= cell_alpha;
      ThreeVector scratchB = mxp;
      scratchB *= coef2;
      scratchA += scratchB;

      p *= coef1;
      p -= scratchA;

      dE_dt_sum += (p*mxH)*Ms_[i]*mesh->Volume(i)*cell_mgamma;
      /// This still needs to be multiplied by MU0 and added
      /// to pE_pt (see below).

      scratchA ^= m;
      mxp *= coef1;
      scratchA += mxp;
      if(do_precess) {
        scratchA -= mxH;
      }

      scratchA *= cell_mgamma;
      dm_dt_[i] = scratchA;
`

**Comment by ilemesh** [comment]
I might be wrong, but applying a similar logic for a Zhang-Li torque (see the derivation for the Slonczewski torque above), it seems that one should change xi-alpha to alpha-xi in the nonadiabatic coefficient of the Zhang-Li torque. The paper [Najafi, Massoud, et al. "Proposal for a standard problem for micromagnetic simulations including spin-transfer torque." Journal of Applied Physics 105.11 (2009): 113914.] also supports my conclusion. Of course, I assume that in the LLG equation one has minus signs: -u*dm/dx-xi*u*[m x dm/dx]. 

As a side note, can anyone explain me why do we have a factor of 2 in the denominator of u in Mumax? (it can be found both in the mumax paper and in the prefactor variable of the file zhangli2.cu). I find no such term neither in the paper above, nor in the original paper by Zhang-Li [Zhang, S., and Z. Li. "Roles of nonequilibrium conduction electrons on the magnetization dynamics of ferromagnets." Physical Review Letters 93.12 (2004): 127204.]

**Comment by barnex** [comment]
addressed by #121
